File: DSA/VaniAgrawal_590015846/Day_55_Question_1.py
# Implement prefix sum tracking with hash map storage to efficiently find longest subarrays meeting exact sum requirements.
arr = list(map(int, input("Enter array elements separated by space: ").split()))
k = int(input("Enter value of k: "))

prefix_sum = 0
sum_map = {}
max_len = 0

for i in range(len(arr)):
    prefix_sum += arr[i]
    if prefix_sum == k:
        max_len = i + 1
    if (prefix_sum - k) in sum_map:
        max_len = max(max_len, i - sum_map[prefix_sum - k])
    if prefix_sum not in sum_map:
        sum_map[prefix_sum] = i

print("Length of longest subarray:", max_len)

File: DSA/PriyankaSingh_590016827/day55question1.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100000

typedef struct {
    int sum;
    int index;
} HashNode;

int findIndex(HashNode hash[], int size, int sum) {
    for (int i = 0; i < size; i++)
        if (hash[i].sum == sum) return hash[i].index;
    return -1;
}

int main() {
    int n, k;
    printf("Enter number of elements: ");
    scanf("%d", &n);
    int arr[n];
    printf("Enter array elements:\n");
    for (int i = 0; i < n; i++) scanf("%d", &arr[i]);
    printf("Enter k: ");
    scanf("%d", &k);

    HashNode hash[MAX];
    int hashSize = 0;
    int sum = 0, maxLen = 0;

    for (int i = 0; i < n; i++) {
        sum += arr[i];
        if (sum == k) maxLen = i + 1;

        int idx = findIndex(hash, hashSize, sum - k);
        if (idx != -1) {
            if (i - idx > maxLen) maxLen = i - idx;
        }

        if (findIndex(hash, hashSize, sum) == -1) {
            hash[hashSize].sum = sum;
            hash[hashSize].index = i;
            hashSize++;
        }
    }

    printf("Longest subarray length: %d\n", maxLen);
    return 0;
}
